mesa/visualization/command_console.py

```diff
@@ -0,0 +1,408 @@
+        self._old_stderr = sys.stderr
+        sys.stdout = self.stdout
+        sys.stderr = self.stderr
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Restore the original stdout and stderr when exiting the context."""
+        sys.stdout = self._old_stdout
+        sys.stderr = self._old_stderr
+
+    def get_output(self):
+        """Retrieve and clear the captured output.
+
+        Returns:
+            tuple: A pair of strings (stdout_output, stderr_output)
+        """
+        output = self.stdout.getvalue()
+        error = self.stderr.getvalue()
+        self.stdout.seek(0)
+        self.stdout.truncate(0)
+        self.stderr.seek(0)
+        self.stderr.truncate(0)
+        return output, error
+
+
+class InteractiveConsole(code.InteractiveConsole):
+    """A custom interactive Python console with output capturing capabilities.
+
+    This class extends code.InteractiveConsole to provide output capturing functionality
+    when executing Python code interactively.
+
+    Args:
+        locals_dict (dict, optional): Dictionary of local variables. Defaults to None.
+    """
+
+    def __init__(self, locals_dict=None):
+        """Initialize the InteractiveConsole with the provided locals dictionary."""
+        super().__init__(locals=locals_dict or {})
+        self.capturer = CaptureOutput()
+
+    def push(self, line):
+        """Push a line to the command interpreter and execute it.
+
+        This method captures the output of the executed command and returns both
+        the 'more' flag and the captured output.
+
+        Args:
+            line (str): The line of code to be executed.
+
+        Returns:
+            tuple: A tuple containing:
+                - more (bool): Flag indicating if more input is needed
+                - str: The captured output from executing the command
+        """
+        with self.capturer:
+            more = super().push(line)
+        return more, self.capturer.get_output()
+
+
+class ConsoleManager:
+    """A console manager for executing Python code interactively.
+
+    This class provides functionality to execute Python code in an interactive console environment,
+    maintain command history, and handle multi-line code blocks.
+
+    Attributes:
+        locals_dict (dict): Dictionary containing local variables available to the console
+        console (InteractiveConsole): Python's interactive console instance
+        buffer (list): Buffer for storing multi-line code blocks
+        history (list[ConsoleEntry]): List of console entries containing commands and their outputs
+    Methods:
+        execute_code(code_line: str, set_input_text: Callable[[str], None]) -> None:
+            Executes the provided code line and updates console history
+        clear_console() -> None:
+            Clears the console history and resets console state
+        get_entries() -> list[ConsoleEntry]:
+            Returns the list of console entries
```

Comment:
```suggestion
```

there is no need to document the methods in the class level docstring. If done this way, they are rendered twice on read the docs